File: Algorithm/src/backtracking/LeetCode79.java
package backtracking;

public class LeetCode79 {


    int[]dx = {-1,1,0,0};
    int[]dy = {0,0,-1,1};


    /**
     * @param board 棋盘
     * @param word 查找的单词
     * @return
     */

    public boolean exist(char[][] board, String word) {

        for(int i = 0; i < board.length; i++){
            for (int j = 0; j < board[0].length; j++){
                if (backtrack(board,word,i,j,0)){
                    return true;
                }
            }
        }
        return false;
    }

    /**
     * @param board 棋盘
     * @param word 查找的单词
     * @param x 棋盘初始位置
     * @param y 棋盘初始位置
     * @return
     */
    private boolean backtrack(char[][]board,String word, int x, int y, int index){
        if (word.length() == index){
            return true;
        }

        if(x < 0 || x >= board.length || y < 0 || y >= board[0].length
                || word.charAt(index) != board[x][y]
        ){
            return false;
        }

        // 查找邻居
        boolean ret = false;
        board[x][y] = '*';
        for(int i  = 0; i < 4; i++){
            if (backtrack(board,word,x+dx[i],y+dy[i],index+1)){
                ret = true;
                break;
            }
        }
        board[x][y] = word.charAt(index);
        return  ret;
    }
}

File: search/DFS/LeetCode79.cc
#include<vector>
#include<unordered_map>
#include<iostream>
using namespace std;

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        if(board.empty()){
            return false;
        }
        int m = board.size(), n = board[0].size();
        vector<vector<bool>> visited(m,vector<bool>(n,false));
        bool find = false;
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                backtracking(i,j,board,word,find,visited,0);
            }
        }
        return find;
    }

    /**
     * @brief 
     * 
     * @param i 
     * @param j 
     * @param board 棋盘大小
     * @param word 带搜索的单词
     * @param find 
     * @param visited 用来标记该位置是否被访问过
     * @param pos  单词位置
     */
    void backtracking(int i, int j, vector<vector<char>>& board,string& word, bool& find, vector<vector<bool>> visited, int pos){
        //判断是否越界,如果越界直接返回
        if(i < 0 || i >= board.size() || j < 0 || j >= board[0].size()){
            return;
        }
        if(visited[i][j] || find ||  board[i][j] != word[pos]){
            return;
        }
        if(pos == word.size()-1){
            find = true;
            return;
        }
        visited[i][j] = true; // 该位置已经被遍历过,修改当前节点状态
        //递归子节点,遍历四个方向
        backtracking(i+1,j,board,word,find,visited,pos+1);
        backtracking(i-1,j,board,word,find,visited,pos+1);
        backtracking(i,j+1,board,word,find,visited,pos+1);
        backtracking(i,j-1,board,word,find,visited,pos+1);
        visited[i][j] = false; 
    }
};